images/cfw/opt/x1plus/lib/python/x1plus/services/x1plusd/polar_cloud.py

```diff
@@ -0,0 +1,274 @@
+"""
+Module to allow printing using Polar Cloud service.
+"""
+
+import asyncio
+import logging
+import os
+import subprocess
+import socketio
+from Crypto.PublicKey import RSA
+from Crypto.Cipher import PKCS1_OAEP
+from base64 import b64encode
+from typing import Dict
+
+
+# from x1plusd.dbus import
+import x1plus
+from x1plus.utils import get_MAC, get_IP, serial_number, is_emulating
+
+logger = logging.getLogger(__name__)
+
+
+class PolarPrintService:
+    def __init__(self, settings):
+        self.polar_sn = 0
+        # Todo: VERY IMPORTANT!! The public and private keys MUST be moved to
+        # non-volatile memory before release.
+        self.public_key = ""
+        self.private_key = ""
+        self.connected = False
+        self.mac = ""
+        self.pin = ""
+        self.username = ""
+        self.server_url = ""
+        self.socket = None
+        self.ip = ""
+        # Todo: Fix two "on" fn calls below. Also, start communicating with dbus.
+        self.polar_settings = settings
+        # self.polar_settings.on("polarprint.enabled", self.sync_startstop())
+        # self.polar_settings.on("self.pin", self.set_pin())
+        self.socket = None
+        self.connected = False  # We might not need this, but here for now.
+        if is_emulating():
+            # I need to use actual account creds to connect, so we're using .env
+            # for testing, until there's an interface.
+            # dotenv isn't installed, so just open the .env file and parse it.
+            # This means that .env file must formatted correctly, with var names
+            # `username`, `pin`, and `server_url`.
+            import inspect
+
+            env_dir = os.path.dirname(inspect.getfile(inspect.currentframe()))
+            with open(os.path.join(env_dir, ".env")) as env:
+                for line in env:
+                    k, v = line.split("=")
+                    setattr(self, k, v.strip())
+
+    async def begin(self):
```

Comment:
We've been calling the top-level threadlet for each module `task`, but I guess this is probably not actually going to be the top-level threadlet: this appears to connect once on printer startup, but does not attempt to repeatedly connect.  Ok for a WIP, I think.

Reply:
Ugh, sorry. I realized that _after_ I shipped. It's actually the first thing I did after I did this PR, when I made PolarPrintService inherit from dbus, so it's planned for the next, or probably the subsequent, PR.